### PR TITLE
more mypy v1.1.1 fixes

### PR DESCRIPTION
**Pull request by williballenthin**
fixes:
  - mypy v1.1.1 for newly merged code
  - more thorough deserialization of result document, as identified by mypy
  - elf typing
  - elf test case failures

### Checklist

- [x] No CHANGELOG update needed

- [x] No new tests needed

- [x] No documentation update needed

### CODE REVIEW

**Comment by williballenthin** [comment]
image

**Comment by williballenthin** [comment]
despite the progress in #1396, there's still a lot of work left to do to deserialize a result document into capa-native object instances. the mypy checks would have caught this, but they were broken when we merged the PR. working on this now.

File: capa/features/extractors/elf.py [hunk 1]
```diff
@@ -773,38 +813,31 @@ def guess_os_from_needed_dependencies(elf) -> Optional[OS]:
...
-    strtab_buf = symtab_buf = None
-
-    for shdr in elf.section_headers:
-        if shdr.type == SHT_STRTAB:
-            strtab_buf, strtab_sz= shdr.buf, shdr.size
-
-        elif shdr.type == SHT_SYMTAB:
-            symtab_buf, symtab_entsize, symtab_sz = shdr.buf, shdr.entsize, shdr.size
-
-    if None in (strtab_buf, symtab_buf):
```
[the remark this answers is not in the document]
Comment (williballenthin) [reply capa/features/extractors/elf.py:788]: note i've also done a little refactoring to make use of existing code to parse section headers and their associated data. no major changes to your logic.

**Comment by williballenthin** [comment]
let's ensure tests pass, and if they do, @mr-tz should consider this for code review and merge, so we can get master green again.

File: capa/features/extractors/elf.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yelhamer) [reply capa/features/extractors/elf.py:788]: Oh, I missed using sh_link to get strtab's index, that's pretty nifty!

I also really like the refactoring you did, specifically, adding symtab() to the ELF section, I think that's much better than my original solution (having it inside the guess_os_from_symtab() function).

Thank you!

**Comment by mr-tz** [comment]
PyInstaller fail appears unrelated (GitHub issue?).

**Comment by williballenthin** [comment]
https://github.com/actions/runner-images/issues/6002

the ubuntu-18.04 image has been deprecated

**Comment by williballenthin** [comment]
i think we were using the oldest ubuntu possible so that the resulting builds would work with the widest range of commonly installed dependencies. by updating to ubuntu 20.04, we may fail to run on other older linuxes (ubuntu 18.04, 18.10, etc.). lets try this out, and if we face pushback, we can find another old linux image.

**Pull request merged**